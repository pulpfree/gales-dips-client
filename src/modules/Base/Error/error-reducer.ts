/**
 * Error Reducer
 *
 * Got most of this from link below
 * @see https://kentcdodds.com/blog/how-to-test-custom-react-hooks?ck_subscriber_id=635632627
 */

import React from 'react'

const CLEAR = 'CLEAR'
const DISMISS = 'DISMISS'
const SEND = 'SEND'

export type ErrorItem = {
  id: string
  message: string
  title?: string
}

export type ErrorState = ErrorItem[]

export type Action =
  | {
      type: 'SEND'
      payload: ErrorItem
    }
  | {
      type: 'DISMISS'
      payload: string
    }
  | {
      type: 'CLEAR'
    }

function errorReducer(state: ErrorState, action: Action): ErrorState {
  switch (action.type) {
    case SEND:
      return [...state, action.payload]

    case DISMISS:
      return state.filter((error) => error.id !== action.payload)

    case CLEAR:
      return []

    default:
      throw new Error('Unhandled action type')
  }
}

export const initialState = {
  errors: [],
}

export const useError = (initialState: ErrorState) => {
  const [state, dispatch] = React.useReducer(errorReducer, initialState)

  const send = React.useCallback(
    (error: ErrorItem) => dispatch({ type: SEND, payload: error }),
    []
  )
  const dismiss = React.useCallback(
    (error: ErrorItem) => dispatch({ type: DISMISS, payload: error.id }),
    []
  )
  const clear = React.useCallback(() => dispatch({ type: CLEAR }), [])

  return { clear, dismiss, send, errors: [...state] }
}
