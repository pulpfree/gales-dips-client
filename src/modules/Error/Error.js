import React from 'react'
import PropTypes from 'prop-types'

import classNames from 'classnames'

import { withStyles } from '@material-ui/core/styles'
import grey from '@material-ui/core/colors/grey'
import Icon from '@material-ui/core/Icon'


const Error = ({ message, classes, type, onClick }) => {
  type = type || 'warning'
  return (
    <div
        className={classNames(classes.base, classes[type])}
        onClick={onClick}
    >
      {message}
      <Icon
          className={classes.icon}
          color="secondary"
      >close</Icon>
    </div>
  )
}

Error.propTypes = {
  classes:  PropTypes.object.isRequired,
  message:  PropTypes.string.isRequired,
  onClick:  PropTypes.func.isRequired,
  type:     PropTypes.string,
}

const styles = theme => ({
  base: {
    borderRadius:     4,
    position: 'relative',
    margin:           theme.spacing.unit,
    marginBottom:     0,
    padding:          theme.spacing.unit * 1.2,
    paddingLeft:      theme.spacing.unit * 1.5,
    paddingRight:     theme.spacing.unit * 1.5,
    color:            grey[800],
  },
  danger: {
    // backgroundColor:  'rgba(242, 69, 61, 0.3)',
    backgroundColor:  '#FBC7C6',
  },
  success: {
    backgroundColor: 'rgba(80, 174, 84, 0.3)',
  },
  info: {
    backgroundColor: 'rgba(31, 188, 211, 0.3)',
  },
  warning: {
    backgroundColor: 'rgba(254, 234, 78, 0.3)',
  },
  icon: {
    position: 'absolute',
    right: 10,
    top: 7,
  },
})

export default withStyles(styles)(Error)
