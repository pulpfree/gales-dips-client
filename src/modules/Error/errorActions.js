export const ERROR_SEND = 'ERROR_SEND'
export const ERROR_DISMISS = 'ERROR_DISMISS'
export const ERROR_CLEAR = 'ERROR_CLEAR'

/**
 * Publish an error alert,
 */
export function errorSend(error) {
  if (!error.id) {
    error.id = new Date().getTime()
  }
  /*return dispatch => {
    dispatch({ type: ERROR_SEND, payload: error })
  }*/
  // console.log('error in errorSend: ', error)
  return { type: ERROR_SEND, payload: error}
}

/**
 * Dismiss an error by the given id.
 */
export function errorDismiss(id) {
  console.log('errorDismiss: ', id)
  return { type: ERROR_DISMISS, payload: id }
}

/**
 * Clear all errors
 */
export function errorClear() {
  return { type: ERROR_CLEAR }
}