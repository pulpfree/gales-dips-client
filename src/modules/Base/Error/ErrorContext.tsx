import React from 'react'

import { ErrorItem, useError } from './error-reducer'

type State = ErrorItem[]
type Dispatch = {
  clear: () => void
  dismiss: (error: ErrorItem) => void
  send: (error: ErrorItem) => void
  errors: ErrorItem[]
}

type ErrorProviderProps = {
  children: React.ReactNode
}

const ErrorStateContext = React.createContext<State | undefined>(undefined)
const ErrorDispatchContext = React.createContext<Dispatch | undefined>(
  undefined
)

export const ErrorProvider = ({ children }: ErrorProviderProps) => {
  const userHandler: Dispatch = useError([])

  return (
    <ErrorStateContext.Provider value={userHandler.errors}>
      <ErrorDispatchContext.Provider value={userHandler}>
        {children}
      </ErrorDispatchContext.Provider>
    </ErrorStateContext.Provider>
  )
}
