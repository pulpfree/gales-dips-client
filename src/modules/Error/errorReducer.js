import { ERROR_SEND, ERROR_DISMISS, ERROR_CLEAR } from './errorActions'

export default function errorReducer(state = [], action) {

  if (!action || !action.type) return state

  switch (action.type) {
  case ERROR_SEND:
    console.log('state: ', state)
    return [action.payload, ...state]

  case ERROR_DISMISS:
    return state.filter(error => error.id !== action.payload)

  case ERROR_CLEAR:
    return []

  default:
    return state
  }
}
