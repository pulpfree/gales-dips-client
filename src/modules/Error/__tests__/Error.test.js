/* eslint no-undef: "off" */

import React from 'react'
import { shallow } from 'enzyme'

import Error from '../Error'

describe('Error', () => {
  let mountedC
  let props
  const component = () => {
    if (!mountedC) {
      mountedC = shallow(<Error {...props} />)
    }
    return mountedC
  }

  beforeEach(() => {
    props = {
      onClick: jest.fn(),
      message: '',
    }
    mountedC = undefined
  })

  it('matches snapshot', () => {
    expect(component()).toMatchSnapshot()
  })
})
