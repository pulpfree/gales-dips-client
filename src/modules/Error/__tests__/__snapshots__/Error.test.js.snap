// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error matches snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "Error-base-1 Error-warning-5",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "aria-hidden": "true",
          "class": "material-icons MuiIcon-root-7 MuiIcon-colorSecondary-9 Error-icon-6",
        },
        "children": Array [
          Object {
            "data": "close",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "aria-hidden": undefined,
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "aria-hidden": undefined,
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
